[Debug] Added "Seach error on Google" button to runtime error page

Lisennk opened the pull request:
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| License | MIT |

It's useful feature because we often google error texts and this button will reduce clicks to do it.

So now runtime error page looks like this:

![snedbs5](https://cloud.githubusercontent.com/assets/243674/17511770/d440fd82-5e24-11e6-9d59-451f81c55cdb.png)

Lisennk committed:
Added "Seach error on Google" button to runtime error page

ro0NL commented on src/Symfony/Component/Debug/ExceptionHandler.php:
```diff
@@ -397,6 +415,12 @@ private function formatArgs(array $args)
+
+    private function formatGoogleLink($query)
+    {
+        return '<a href="//google.com/search?q=' . urlencode(strip_tags($query)) . '" target="_blank" rel="noopener noreferrer" title="Search error on Google"><img src="//google.com/favicon.ico" alt="Search error on Google"></a>';
```
This is nowadays an anti-pattern, we should go with https whenever possible.

Lisennk committed:
Moving to HTTPS due to @ro0NL comment

ro0NL commented:
[ref](https://www.paulirish.com/2010/the-protocol-relative-url/)

Lisennk committed:
Properly merge into latest commit

Properly merge into latest commit to prevent deleting code that doesn't relate to the feature

Taras-S commented:
ro0NL, done. Fixed Protocol-relative URL and removed accidental erasure of latest code (yesterday i forget about it)

wouterj commented:
I think a "search on google" link instead of the icon would be better for both the consistency of the design and the UX.

ro0NL commented:
We also use _their_ favicon currently.. (`//google.com/favicon.ico` which is still protocol relative ;-)) maybe base64 it or go with text indeed, it looks pretty though :angel:

Taras-S commented:
Ok, thanks,  i will replace it to link soon

Lisennk committed:
Replaced Google icon with "Search on Google" text

Taras-S commented:
It's done now: icon replaced by text.

nicolas-grekas commented:
We have another error page on Symfony apps (the one with the little ghost).
I guess you're no using Symfony standard edition?
Should we also add this search link to the Symfony standard edition error page?
@wouterj @javiereguiluz please advise :)

Lisennk commented:
actually im using Laravel right now, and Laravel uses this component to display errors. But if you think that we also need to add "search on google" link to other error page, i can do it. So, we need it?

Lisennk commented:
So, where this page with ghost is located?

nicolas-grekas commented:
The template is `src/Symfony/Bundle/TwigBundle/Resources/views/Exception/exception.html.twig`

Lisennk commented:
OK, working on it right now.

Lisennk committed:
Add "Google Search" to Twig exception pae

Lisennk commented:
@ro0NL @nicolas-grekas @wouterj, as you wish, i added this link to twig exception page too.

sstok commented on src/Symfony/Bundle/TwigBundle/Resources/views/Exception/exception.html.twig:
```diff
@@ -1,7 +1,7 @@
     <div class="block-exception-detected clear-fix">
         <div class="support">
-            <a href="http://symfony.com/support">Need support?</a>
+            <a href="http://symfony.com/support">Support</a> | <a href="https://google.com/search?q={{ exception.message|nl2br|format_file_from_text }}" target="_blank">Google Search</a>
```
This could use the `rel="noopener noreferrer"` as well 👍

Lisennk committed:
Security fix

Lisennk commented:
@sstok, fixed it

Lisennk commented:
Build with last commit fails on PHP 7.1. Any suggestions?

nicolas-grekas commented on src/Symfony/Bundle/TwigBundle/Resources/views/Exception/exception.html.twig:
```diff
@@ -1,7 +1,7 @@
     <div class="block-exception-detected clear-fix">
         <div class="support">
-            <a href="http://symfony.com/support">Need support?</a>
+            <a href="http://symfony.com/support">Support</a> | <a href="https://google.com/search?q={{ exception.message|nl2br|format_file_from_text }}" rel="noopener noreferrer" target="_blank">Google Search</a>
```
not sure about `|nl2br|format_file_from_text`
would it work better with `|url_encode`?

nicolas-grekas replied:
or `|striptags|url_encode`?

wouterj replied:
Indeed, seems like this is a copy/past error. `format_file_from_text` and `nl2br` both generate HTML tags, which surely is something we don't want here.

I don't think tags exists in the error message, so `striptags` doesn't make sense as well.

Btw, I think we need a customized filter. E.g. error messages often show app specific stuff (like file paths, service names, etc.) those should be stripped imo as they will often result in "0 results found" on Google.

Lisennk replied:
Yes, maybe someone,  who knows Symfony better than me, will implement this filter it in future

stof replied:
however, we need ``|escape('url')`` to handle the proper url encoding for usage in the URL. Otherwise, special chars will cause issues

stof commented on src/Symfony/Component/Debug/ExceptionHandler.php:
```diff
@@ -399,4 +404,12 @@ private function escapeHtml($str)
+     */
+    private function formatGoogleLink($query)
+    {
+        return '<a href="https://google.com/search?q=' . urlencode(strip_tags($query)) . '" target="_blank" rel="noopener noreferrer" title="Search error on Google">Search on Google</a>';
```
no need to strip tags IMO. However, you need to apply HTML escaping (after the URL encoding), as this URL gets displayed in some HTML.

Lisennk replied:
@stof, you mean `$this->escapeHtml(urlencode(strip_tags($query)))`?

shakaran replied:
@stof my 2 cent, probably the title in "Search error on Google" and text link "Search on Google" should support l18n, using hardcoded english words probably is not a good idea.

nicolas-grekas commented:
@Lisennk does this really work? I mean: you said earlier that long exception messages give no result on Google because they are too specific.

javiereguiluz commented:
@Lisennk I thank you for all the work you did here ... but I'm not sure about this feature. I'm 👎 for now because of:

* Any person that may use this feature is an advanced developer who probably has their preferred way to look for help in Google or other search engine.
* To be really useful ... this feature should do an advanced processing of the exception message: remove custom file names and paths (e.g. `/projects/javier/...`), remove common words (the, a, an, etc.), be careful with punctuation characters (`+` and `-` have special meaning in Google), be careful to not use too many words in the search query, etc.

Lisennk commented:
If i will implement second item of your answer, you will be :+1:? Anyway, thanks for your time.

phansys commented:
Regardless Google™ is the industry lead on search engines (and many, many other scopes), could we use more abstract / decoupled approach in order to allow this feature to be extended or configured in a future? I mean, maybe someone prefers the results from another search engine (that maybe consumes the same Google's SERPs and apply some custom layout on top of them :trollface:).

phansys commented:
BTW, since the whole Symfony documentation is available at symfony.com and the only official support channel is stackoverflow.com, maybe the search should be restricted with `site:` operator to match only results from these domains, in order to act as an "internal" search engine... although, thinking it again, both of these sites already have its own search engines:
* https://symfony.com/search?q={term}
* https://stackoverflow.com/search?q={term}

nicolas-grekas commented:
But as @Lisennk said, he is from a Laravel background, thus Symfony doc/SO might not be the most universal here. 

Instead of coding some complicated cleaning logic, "select+right click > search on Google" is in all browsers. May be the most universal to me...

nicolas-grekas commented:
Closing as discussed, anyway, thanks for trying @Lisennk!

Lisennk commented:
You are welcome :)
